eclipse-che/che-theia
# Make Git:clone consistent with cloning of projects by workspace plugin

Signed-off-by: Vitaliy Gulyy <vgulyy@redhat.com>



### What does this PR do?
- Handles cloning of a project with Git: clone command, displays a notification to solve SSH issues
- Adds all SSH keys to authentication agent on workspace startup, asks for SSH passphrase if needed

### Screenshot/screencast of this PR

![Screenshot from 2021-04-08 14-27-51](https://user-images.githubusercontent.com/1655894/114029892-a9fc8080-9882-11eb-980d-fe5575fc02a8.png)


### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/19133
https://github.com/eclipse/che/issues/19345

### How to test this PR?


Pay attention, that this section describes how to test the PR in single root workspace.
In multi root workspace, the project is being cloned, but is not appeared in the explorer.
This is known problem https://github.com/eclipse/che/issues/19544, with a corresponding PR https://github.com/eclipse-che/che-theia/pull/1076 to fix it.

1. Create a workspace from the devfile
```
---
apiVersion: 1.0.0
metadata:
  name: test-git-clone
attributes:
  multiRoot: 'off'
components:
  - type: cheEditor
    alias: che-theia
    reference: https://raw.githubusercontent.com/chepullreq4/pr-check-files/master/che-theia/pr-1068/che_theia_meta.yaml
    memoryLimit: 2Gi
```

2. In Che-Theia open command palette, run `Git: clone` command and clone any ssh project, e.g. `git@github.com:eclipse/che-theia.git`

You will be notified, that there is a missing ssh key. A notification with proposal to configure ssh should appear.

To test https://github.com/eclipse/che/issues/19345 you need to:
- upload encrypted ssh key with `SSH: Upload private key...`. You will be asked for ssh passphrase.
- restart the workspace. You will be asked for the ssh passphrase again (just because the ssh key is added to ssh authentication agent on the start)
- clone any git repository with ssh uri to test

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)


### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### Happy Path Channel

HAPPY_PATH_CHANNEL=next
